[docs] Redesign

stevhliu opened the pull request:
The main goal of this PR is to redesign the Transformers docs to:

1. Be more developer-friendly.
2. Improve navigation by replacing the existing structure with a more organic one that scales naturally instead of forcing content into the 4 current predefined sections.
3. Create a more unified docs experience by integrating content rather than adding it on.

This PR proposes a potential structure for achieving 2 and 3. Once the structure is in place, each doc will be rewritten to achieve 1.

If you're interested in more details about the redesign's motivation, please read this [blog post](https://www.stevhliu.com/2024/making-sense-of-this-mess). If you want more details about 1, 2, and 3, please read this [post](https://www.stevhliu.com/2024/unraveling-the-mess) and this [one](https://www.stevhliu.com/2024/diagrams) too.

All feedback, alternative structures, and comments are welcomed! Thanks 🙂

stevhliu committed:
toctree

stevhliu committed:
not-doctested.txt

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_31757). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

stevhliu committed:
collapse sections

gante commented on docs/source/en/_toctree.yml:
```diff
@@ -1,144 +1,158 @@
+      title: Getting the most out of LLMs
+    - local: perplexity
+      title: Perplexity of fixed-length models
+  - title: Chat models
```
perhaps move this down a level (i.e. inside LLMs)? a Chat model is a subtype of LLM

gante commented on docs/source/en/_toctree.yml:
```diff
@@ -1,144 +1,158 @@
+      title: Pipelines for webserver inference
+    - local: add_new_pipeline
+      title: How to add a pipeline to 🤗 Transformers?
+  - title: LLMs
```
A type of model that's becoming increasingly common are VLMs: they are the same as LLMs, but also accept image inputs.

Would it make sense to call this section "LLMs and VLMs"?

stevhliu replied:
For sure! Let's rename the section when we have some VLM-specific docs?

gante reviewed (approved):
Like! 👍

ydshieh commented on docs/source/en/_toctree.yml:
```diff
@@ -1,144 +1,158 @@
-  title: Tutorials
-- sections:
-  - isExpanded: false
+- title: Pretrained models
```
Maybe we can come up with a better title here, as it's not only about models.

stevhliu replied:
Changed to "Base classes"

ydshieh commented on docs/source/en/_toctree.yml:
```diff
@@ -1,144 +1,158 @@
-    - local: tasks/mask_generation
-      title: Mask Generation
-    - local: tasks/knowledge_distillation_for_image_classification
-      title: Knowledge Distillation for Computer Vision
-    title: Computer Vision
-  - isExpanded: false
+    - local: serialization
+      title: Export to ONNX
+    - local: tflite
+      title: Export to TFLite
+    - local: torchscript
+      title: Export to TorchScript
```
I think this would be better in a top level instead of under `title: Pretrained models`

ydshieh commented on docs/source/en/_toctree.yml:
```diff
@@ -159,91 +173,94 @@
-  title: Conceptual guides
 - sections:
-  - sections:
+  - isExpanded: false
+    sections:
```
should  ` title: Main Classes` be moved around here?

ydshieh replied:
similarly, move `title: Text models` etc.

ydshieh commented:
Indeed easier to read ❤️ . But there are a few places need to be moved if I understand correctly?

stevhliu commented on docs/source/en/_toctree.yml:
```diff
@@ -1,144 +1,158 @@
+      title: Export to TFLite
+    - local: torchscript
+      title: Export to TorchScript
+- title: Training
```
Totally! Moved inference ahead of training 🙂

stevhliu committed:
feedback

stevhliu committed:
update

stevhliu committed:
rewrite get started sections

stevhliu committed:
fixes

stevhliu committed:
fix

stevhliu commented:
I've kicked off the redesign with the "Get Started" section. Feel free to review this section while I start on the next one (Base classes)!

The main changes are:

`index.md`
- A cleaner index page that better describes *what* Transformers is in terms of its features and design. I believe this is more impactful than listing all the tasks you can solve across modalities. The focus shouldn't be on the tasks that you can solve; it should be on the models themselves. By describing the type of models available, I think users will understand that they can use them for their tasks. Having a more holistic description of the library here is more important than focusing on the different tasks/modalities.
- More of a question here, but would it be better to maybe have badges on each model API doc that indicate whether it supports PyTorch/TensorFlow/Flax? Instead of having/maintaining such a long list that clutters up the main landing page, I think it'd be a lot cleaner to have this information on each model page. This way, users can see everything at once on the model page.

`quicktour.md`
- Removed the "vertical" PyTorch/TensorFlow blocks in favor of the "horizontal" ones which I think it cleaner and less overwhelming.
- Removed the big table of tasks available to `Pipeline` with just three code examples. I think this makes it simpler and more approachable. I also removed the `Pipeline` video because it felt very NLP-heavy, but we can add it back if we want to keep it.
- Updated the AutoClass section to also be simpler and faster for users to start. A lot of these details (eg, tensors are outputted before the final activation function, custom model builds) can be explained in more depth in later docs. Also took this opportunity to introduce the `generate` API.
- A better Next steps section directing users to topics of interest.

`installation.md`
- Removed the options for downloading files in favor of just one method to keep it simple, and link to the Download files from the Hub doc for more details.

stevhliu committed:
loading models

stevhliu committed:
fix

stevhliu committed:
customize models

stevhliu committed:
share

stevhliu committed:
fix link

stevhliu committed:
contribute part 1

stevhliu committed:
contribute pt 2

stevhliu committed:
fix toctree

stevhliu commented:
Hi, I'm back with an update! I've wrapped up the technical guides in the Models section. I'll circle back to the more conceptual docs later and also create some visual diagrams in Figma. Next up, I'll start working on the Preprocessors section. 🙂 

The main focus is on how to **load**, **customize**, **share**, and **contribute** a model, basically a one-stop section for all your **general** model docs. The Load and Contribute docs have more significant changes:

`models.md`
- Repurposed to show how to load a model. I start with a quick example of `AutoModelFor.from_pretrained()` so you can immediately get started, and then progressively peel back the layers. From how models and configurations interact, to the AutoClass API, and then model-specific classes. To make it easier to find how to load any model, I also added big models (`device_map="auto"`) and custom models (`trust_remote_code="True"`)to this page. 

`add_new_model.md`
- Updated structure to make the steps more discoverable. Before, many of the steps were hidden in "5.-14. Port BrandNewBert to Transformers" but now what these actual steps are more clear.

younesbelkada committed:
Add new model (#32615)

* v1 - working version

* fix

* fix

* fix

* fix

* rename to correct name

* fix title

* fixup

* rename files

* fix

* add copied from on tests

* rename to `FalconMamba` everywhere and fix bugs

* fix quantization + accelerate

* fix copies

* add `torch.compile` support

* fix tests

* fix tests and add slow tests

* copies on config

* merge the latest changes

* fix tests

* add few lines about instruct

* Apply suggestions from code review

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

* fix

* fix tests

---------

Co-authored-by: Arthur <48595927+ArthurZucker@users.noreply.github.com>

qgallouedec committed:
"to be not" -> "not to be" (#32636)

* "to be not" -> "not to be"

* Update sam.md

* Update trainer.py

* Update modeling_utils.py

* Update test_modeling_utils.py

* Update test_modeling_utils.py

stevhliu committed:
tokenization pt 1

stevhliu committed:
fix hfoption tag

stevhliu committed:
tokenization pt. 2

stevhliu commented:
Finished the first draft of the [Tokenizers](https://moon-ci-docs.huggingface.co/docs/transformers/pr_31757/en/fast_tokenizers) doc, and I'm pretty excited that it reduces "content creep" from 6 different docs to just 1! 😎

stevhliu committed:
image processor

stevhliu committed:
fix toctree

stevhliu committed:
backbones

stevhliu committed:
feature extractor

stevhliu committed:
fix file name

stevhliu committed:
processor

stevhliu committed:
update not-doctested

stevhliu commented:
The first draft of the practical guides in the **base classes** section is finished now! Please feel free to check it out and leave any comments or feedback (not sure why the feature extractor and processor docs aren't showing in the preview at the moment) 😄 

I'll start working on the **inference** section after I review the first draft.

stevhliu committed:
update

stevhliu committed:
make style

stevhliu committed:
fix toctree

stevhliu committed:
revision

stevhliu committed:
make fixup

stevhliu committed:
fix toctree

stevhliu committed:
fix

stevhliu committed:
make style

stevhliu committed:
fix hfoption tag

stevhliu committed:
pipeline

stevhliu committed:
pipeline gradio

stevhliu committed:
pipeline web server

stevhliu committed:
add pipeline

stevhliu committed:
fix toctree

stevhliu committed:
not-doctested

stevhliu committed:
prompting

stevhliu committed:
llm optims

stevhliu committed:
fix toctree

stevhliu committed:
fixes

stevhliu committed:
cache

stevhliu committed:
text generation

stevhliu committed:
fix

stevhliu committed:
chat pipeline

stevhliu committed:
chat stuff

stevhliu committed:
xla

stevhliu committed:
torch.compile

stevhliu committed:
cpu inference

stevhliu committed:
toctree

stevhliu committed:
gpu inference

stevhliu committed:
agents and tools

stevhliu committed:
gguf/tiktoken

stevhliu committed:
finetune

stevhliu committed:
toctree

stevhliu committed:
trainer

stevhliu committed:
trainer pt 2

stevhliu committed:
optims

stevhliu committed:
optimizers

stevhliu committed:
accelerate

stevhliu committed:
parallelism

stevhliu committed:
fsdp

stevhliu committed:
update

stevhliu committed:
distributed cpu

stevhliu committed:
hardware training

stevhliu committed:
gpu training

stevhliu committed:
gpu training 2

stevhliu committed:
peft

stevhliu committed:
distrib debug

stevhliu committed:
deepspeed 1

stevhliu committed:
deepspeed 2

stevhliu committed:
chat toctree

stevhliu committed:
quant pt 1

stevhliu committed:
quant pt 2

stevhliu committed:
fix toctree

stevhliu committed:
fix

stevhliu committed:
fix

stevhliu committed:
quant pt 3

stevhliu committed:
quant pt 4

stevhliu committed:
serialization

stevhliu committed:
torchscript

stevhliu committed:
scripts

stevhliu committed:
tpu

stevhliu committed:
review

stevhliu committed:
model addition timeline

stevhliu committed:
modular

stevhliu committed:
more reviews

stevhliu committed:
reviews

stevhliu committed:
fix toctree

stevhliu committed:
reviews reviews

stevhliu committed:
continue reviews

stevhliu committed:
more reviews

stevhliu committed:
modular transformers

stevhliu committed:
more review

stevhliu committed:
zamba2

stevhliu committed:
fix

stevhliu committed:
all frameworks

stevhliu committed:
pytorch

ArthurZucker commented on docs/source/en/_toctree.yml:
```diff
@@ -1,279 +1,303 @@
   - local: quicktour
-    title: Quick tour
+    title: Quickstart
   - local: installation
```
Would probably put installation first 😄

ArthurZucker commented on docs/source/en/_toctree.yml:
```diff
@@ -1,279 +1,303 @@
-  - isExpanded: false
+    - local: fast_tokenizers
+      title: Tokenizers
+    - local: tiktoken
+      title: tiktoken
```
Imo this does not have it's place here

stevhliu replied:
Moved into the tokenizers [doc](https://moon-ci-docs.huggingface.co/docs/transformers/pr_31757/en/fast_tokenizers)

ArthurZucker commented on docs/source/en/_toctree.yml:
```diff
@@ -1,279 +1,303 @@
-    title: Audio
-  - isExpanded: false
+    - local: models
+      title: Loading
```
could be clearer

ArthurZucker commented on docs/source/en/_toctree.yml:
```diff
@@ -1,279 +1,303 @@
+    - local: model_sharing
+      title: Sharing
+    - local: add_new_model
+      title: Adding a new model
```
```suggestion
      title: Adding a new model to transformers
```

ArthurZucker commented on docs/source/en/_toctree.yml:
```diff
@@ -1,279 +1,303 @@
+    - local: add_new_model
+      title: Adding a new model
+    - local: modular_transformers
+      title: Modular Transformers
```
This might be better under adding a model, or close to the philosophy

stevhliu replied:
I think since Adding a new model is already so long, its better to keep Modular Transformers separate. Having them next to each other is ideal so users don't have to skip around too much in the sidebar!

ArthurZucker commented on docs/source/en/_toctree.yml:
```diff
@@ -1,279 +1,303 @@
+      title: Getting the most out of LLMs
+    - local: perplexity
+      title: Perplexity of fixed-length models
+  - title: Chat
```
Can be detailed imo chat is uclear to me

stevhliu replied:
Updated to Chat with models :)

ArthurZucker commented on docs/source/en/_toctree.yml:
```diff
@@ -1,279 +1,303 @@
+    sections:
+    - local: perf_torch_compile
+      title: torch.compile
+    - local: tf_xla
+      title: XLA
```
would put this last

ArthurZucker commented on docs/source/en/_toctree.yml:
```diff
@@ -1,279 +1,303 @@
+      title: Tools and RAG
+  - title: Optimization
+    sections:
+    - local: perf_torch_compile
```
might need to add a section for vllm / tgi backend support!

stevhliu replied:
For sure, will do in a new and separate PR if thats ok

ArthurZucker commented on docs/source/en/_toctree.yml:
```diff
@@ -1,279 +1,303 @@
+- title: Export to production
+  isExpanded: False
+  sections:
+  - local: serialization
+    title: ONNX
+  - local: tflite
+    title: LiteRT
+  - local: torchscript
+    title: TorchScript
```
executorch missing!

stevhliu replied:
Added based on code [here](https://github.com/huggingface/transformers/blob/62db3e6ed67a74cc1ed1436acd9973915c0a4475/tests/models/llama/test_modeling_llama.py#L799)!

ArthurZucker commented on docs/source/en/add_new_model.md:
```diff
@@ -13,496 +13,308 @@ rendered properly in your Markdown viewer.
-  have to look into the respective `modeling_....py` file.
+- composition > over-abstraction
+- duplicate code isn't always bad if it greatly improves readability and accessibility
+- model files are self-contained and all the necessary model code is found in the `modeling.py` file
```
```suggestion
- model files are self-contained and all the necessary model code is found in the `modeling_mymodel.py` file
```

ArthurZucker commented on docs/source/en/add_new_model.md:
````diff
@@ -13,496 +13,308 @@ rendered properly in your Markdown viewer.
 
-Jupyter notebooks have the advantage that they allow for cell-by-cell execution which can be helpful to better split
-logical components from one another and to have faster debugging cycles as intermediate results can be stored. Also,
-notebooks are often easier to share with other contributors, which might be very helpful if you want to ask the Hugging
-Face team for help. If you are familiar with Jupyter notebooks, we strongly recommend you work with them.
+  ```py
+  # Copied from transformers.models.bert.modeling_bert.BertSelfAttention with Bert->Roberta
+  class RobertaSelfAttention(nn.Module):
````
IMO no longer true as we want to push for modular usage!

ArthurZucker commented on docs/source/en/add_new_model.md:
```diff
@@ -13,496 +13,308 @@ rendered properly in your Markdown viewer.
-It is recommended that you retrieve the output, either by print statements or sub-component functions, of the following
-layers in the following order:
+> [!TIP]
+> Each contributor has a unique style and workflow for porting models to Transformers. It may be helpful to take a look at how [GPT2](https://medium.com/huggingface/from-tensorflow-to-pytorch-265f40ef2a28) and [WMT19](https://huggingface.co/blog/porting-fsmt) were ported.
```
Imo too outdated, would rather link PRs adding Gemma for example

ArthurZucker commented on docs/source/en/add_new_model.md:
```diff
@@ -13,496 +13,308 @@ rendered properly in your Markdown viewer.
+- Don't be shy to ask for help! We are here to support you. 🤗
 
-Input IDs should thereby consists of an array of integers, *e.g.* `input_ids = [0, 4, 4, 3, 2, 4, 1, 7, 19]`
+## Dev environment
```
nice to isolate this!

ArthurZucker commented on docs/source/en/add_new_model.md:
````diff
@@ -13,496 +13,308 @@ rendered properly in your Markdown viewer.
+```py
+model = BrandNewBertModel.load_pretrained_checkpoint("/path/to/checkpoint/")
+input_ids = [0, 4, 5, 2, 3, 7, 9]  # vector of input ids
+original_output = model.predict(input_ids)
````
never seem predict, but IMO we could change this to be BrandNewLlamaModel instead no?

stevhliu replied:
Done, replaced all mentions of `BrandNewBertModel` with `BrandNewLlamaModel`!

ArthurZucker commented on docs/source/en/add_new_model.md:
```diff
@@ -622,339 +425,202 @@ tensor([[-0.0818,  0.2207, -0.0749, -0.0030,  0.0045, -0.1569, -0.1598,  0.0212,
 
-Next, we should add the tokenizer of *brand_new_bert*. Usually, the tokenizer is equivalent to or very similar to an
-already existing tokenizer of 🤗 Transformers.
+## Implement tokenizer
```
IMO we should recommend adding fast tokenizer / asking me or @itazap how to convert to `PreTrainedTokenizerFast`!

ArthurZucker commented on docs/source/en/cache_explanation.md:
wow nice!

ArthurZucker commented on docs/source/en/chat_extras.md:
````diff
@@ -0,0 +1,299 @@
+
+```py
+inputs = tokenizer.apply_chat_template(messages, tools=tools, add_generation_prompt=True, return_dict=True, return_tensors="pt")
+inputs = {k: v.to(model.device) for k, v in inputs.items()}
````
you don't have to do this, normally inputs support `to` directly

ArthurZucker commented on docs/source/en/chat_extras.md:
````diff
@@ -0,0 +1,299 @@
+
+```py
+inputs = tokenizer.apply_chat_template(messages, tools=tools, add_generation_prompt=True, return_dict=True, return_tensors="pt")
+inputs = {k: v.to(model.device) for k, v in inputs.items()}
````
same here and everywhere

ArthurZucker reviewed (commented):
Mostly review adding a new model but nice and refreshing! Huge work, congrats!

SunMarc commented on docs/source/en/quantization/overview.md:
```diff
@@ -14,81 +14,24 @@ rendered properly in your Markdown viewer.
-
-Use the table below to help you decide which quantization method to use.
-
-| Quantization Method                           | On the fly quantization | CPU             | CUDA GPU | ROCm GPU  | Metal (Apple Silicon)              | Intel GPU       | Torch compile() | Bits          | PEFT Fine Tuning | Serializable with 🤗Transformers | 🤗Transformers Support  | Link to library                             |
-|-----------------------------------------------|----------------------|-----------------|----------|-----------|------------------------------------|-----------------|-----------------|---------------|------------------|-----------------------------|-------------------------|---------------------------------------------|
-| [AQLM](./aqlm.md)                             | 🔴                   | 🟢              |     🟢     | 🔴        | 🔴                                 | 🔴              | 🟢              | 1/2         | 🟢               | 🟢                          | 🟢                      | https://github.com/Vahe1994/AQLM            |
-| [AWQ](./awq.md)                               | 🔴                   | 🟢              | 🟢        | 🟢        | 🔴                                 | 🟢              | ?               | 4             | 🟢               | 🟢                          | 🟢                      | https://github.com/casper-hansen/AutoAWQ    |
-| [bitsandbytes](./bitsandbytes.md)             | 🟢                   | 🟡 <sub>1</sub> |     🟢     | 🟡 <sub>1</sub> | 🔴 <sub>2</sub>                    | 🟡 <sub>1</sub> | 🔴 <sub>1</sub> | 4/8         | 🟢               | 🟢                          | 🟢                      | https://github.com/bitsandbytes-foundation/bitsandbytes |
-| [compressed-tensors](./compressed_tensors.md) | 🔴                   | 🟢              |     🟢     | 🟢        | 🔴                                 | 🔴              | 🔴              | 1/8         | 🟢               | 🟢                          | 🟢                      | https://github.com/neuralmagic/compressed-tensors |
-| [EETQ](./eetq.md)                             | 🟢                   | 🔴              | 🟢        | 🔴        | 🔴                                 | 🔴              | ?               | 8             | 🟢               | 🟢                          | 🟢                      | https://github.com/NetEase-FuXi/EETQ        |
-| [GGUF / GGML (llama.cpp)](../gguf.md)         | 🟢                   | 🟢              | 🟢        | 🔴        | 🟢                                 | 🔴              | 🔴              | 1/8         | 🔴               | [See Notes](../gguf.md)     | [See Notes](../gguf.md) | https://github.com/ggerganov/llama.cpp      |
-| [GPTQModel](./gptq.md)                        | 🔴                   | 🟢 <sub>3</sub> | 🟢        | 🟢        | 🟢                                 | 🟢 <sub>4</sub> | 🔴              | 2/3/4/8 | 🟢               | 🟢                          | 🟢                      | https://github.com/ModelCloud/GPTQModel        |
-| [AutoGPTQ](./gptq.md)                         | 🔴                   | 🔴              | 🟢        | 🟢        | 🔴                                 | 🔴              | 🔴              | 2/3/4/8 | 🟢               | 🟢                          | 🟢                      | https://github.com/AutoGPTQ/AutoGPTQ        |
```
I prefer to keep this table if possible !

stevhliu replied:
Keeping for now, but will continue to look for another way to better present this info :)

SunMarc commented on docs/source/en/quantization/torchao.md:
```diff
@@ -9,81 +9,56 @@ specific language governing permissions and limitations under the License.
-# We support int4_weight_only, int8_weight_only and int8_dynamic_activation_int8_weight
-# More examples and documentations for arguments can be found in https://github.com/pytorch/ao/tree/main/torchao/quantization#other-available-quantization-techniques
 quantization_config = TorchAoConfig("int4_weight_only", group_size=128)
-quantized_model = AutoModelForCausalLM.from_pretrained(model_name, torch_dtype="auto", device_map="auto", quantization_config=quantization_config)
-
-tokenizer = AutoTokenizer.from_pretrained(model_name)
-input_text = "What are we having for dinner?"
-input_ids = tokenizer(input_text, return_tensors="pt").to("cuda")
-
-# auto-compile the quantized model with `cache_implementation="static"` to get speedup
-output = quantized_model.generate(**input_ids, max_new_tokens=10, cache_implementation="static")
-print(tokenizer.decode(output[0], skip_special_tokens=True))
```
can we keep the generation part also since it is important to show that with static cache, we get a nice speedup with torchao. We can skip the benchmark.

SunMarc commented on docs/source/en/quantization/torchao.md:
````diff
@@ -9,81 +9,56 @@ specific language governing permissions and limitations under the License.
-# save_to = "{user_id}/llama3-8b-int4wo-128"
-# quantized_model.push_to_hub(save_to, safe_serialization=False)
+quantized_model = torch.compile(quantized_model, mode="max-autotune")
+```
````
We don't have to compile the model for faster generation. As said previously, just using static cache will get you speed-up as we will compile the forward automatically.

SunMarc reviewed (commented):
Left a few comments. Thanks for this huge work !

stevhliu committed:
supported model frameworks

stevhliu committed:
flashattention

gante commented on docs/source/en/llm_tutorial.md:
```diff
@@ -14,266 +14,264 @@ rendered properly in your Markdown viewer.
+</hfoption>
+<hfoption id="chat template">
 
-### Related libraries
```
Has this been moved? At a first glance, I can't find it

(I've been using this section to promote certain related libraries 🤗 )

gante reviewed (commented):
This is a lot of work 👀 

(I've reviewed pages directly related to text generation)

stevhliu committed:
rm check_table

stevhliu committed:
not-doctested.txt

stevhliu committed:
rm check_support_list.py

stevhliu committed:
feedback

stevhliu commented on docs/source/en/_toctree.yml:
```diff
@@ -1,279 +1,303 @@
+      title: The Transformer model family
+    - local: attention
+      title: Attention mechanisms
+  - title: Preprocessors
```
I think the Preprocessor section should already be not expanded by default

stevhliu commented:
Thanks so much for the feedback everyone! If you can take a look at the docs related to the topics you're working on that'd be amazing (but feel free to explore the other docs as well if you want!) 🫶 

Some other notable changes include removing the `check_table` and `check_support_list` scripts because these are now replaced by individual badges on the model API page (see example [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_31757/en/model_doc/bamba) for SDPA/FlashAttention support and [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_31757/en/model_doc/albert) for framework support).

stevhliu committed:
updates/feedback

stevhliu committed:
review

matthewdouglas commented on docs/source/en/quantization/bitsandbytes.md:
```diff
@@ -16,42 +16,24 @@ rendered properly in your Markdown viewer.
+[bitsandbytes](https://github.com/bitsandbytes-foundation/bitsandbytes) features the LLM.int8 and QLoRA quantization to enable accessible large language model inference and training.
 
-To use bitsandbytes, make sure you have the following libraries installed:
+LLM.int8 matrix multiplication, or 8-bit quantization, is based on vector-wise quantization to quantize most of the weights to 8-bits and treating outliers with 16-bit matrix multiplication to reduce their degradative effect on model accuracy.
 
```
```suggestion
[LLM.int8()](https://hf.co/papers/2208.07339) is a quantization method that aims to make large language model inference more accessible without significant degradation. Unlike naive 8-bit quantization, which can result in loss of critical information and accuracy, LLM.int8() dynamically adapts to ensure sensitive components of the computation retain higher precision when needed.
```

matthewdouglas commented on docs/source/en/quantization/bitsandbytes.md:
```diff
@@ -16,42 +16,24 @@ rendered properly in your Markdown viewer.
-</Tip>
-
-Now you can quantize a model by passing a `BitsAndBytesConfig` to [`~PreTrainedModel.from_pretrained`] method. This works for any model in any modality, as long as it supports loading with Accelerate and contains `torch.nn.Linear` layers.
+Quantize a model by passing a [`BitsAndBytesConfig`] to [`~PreTrainedModel.from_pretrained`]. This works for any model in any modality, as long as it supports [Accelerate]https://huggingface.co/docs/accelerate/index() and contains [torch.nn.Linear](https://pytorch.org/docs/stable/generated/torch.nn.Linear.html) layers.
```
```suggestion
Quantize a model by passing a [`BitsAndBytesConfig`] to [`~PreTrainedModel.from_pretrained`]. This works for any model in any modality, as long as it supports [Accelerate](https://huggingface.co/docs/accelerate/index) and contains [torch.nn.Linear](https://pytorch.org/docs/stable/generated/torch.nn.Linear.html) layers.
```

matthewdouglas commented on docs/source/en/quantization/bitsandbytes.md:
````diff
@@ -64,7 +46,7 @@ model_8bit = AutoModelForCausalLM.from_pretrained(
 ```
 
-By default, all the other modules such as `torch.nn.LayerNorm` are converted to `torch.float16`. You can change the data type of these modules with the `torch_dtype` parameter if you want. Setting `torch_dtype="auto"` loads the model in the data type defined in a model's `config.json` file.
+By default, all other modules such as [torch.nn.LayerNorm](https://pytorch.org/docs/stable/generated/torch.nn.LayerNorm.html) are converted to `torch.float16`. You can change the data type of these modules with the `torch_dtype` parameter.. Setting `torch_dtype="auto"` loads the model in the data type defined in a model's `config.json` file.
````
I'm not sure this is accurate. I would think it would follow whatever the torch default dtype is unless `torch_dtype` is set. The bitsandbytes integration does not do anything to a module is not `nn.Linear`.

stevhliu replied:
Clarified that it is set to just the default torch dtype unless explicitly set with `torch_dtype` parameter!

matthewdouglas commented on docs/source/en/quantization/bitsandbytes.md:
```diff
@@ -200,7 +173,7 @@ model_8bit = AutoModelForCausalLM.from_pretrained(
 An "outlier" is a hidden state value greater than a certain threshold, and these values are computed in fp16. While the values are usually normally distributed ([-3.5, 3.5]), this distribution can be very different for large models ([-60, 6] or [6, 60]). 8-bit quantization works well for values ~5, but beyond that, there is a significant performance penalty. A good default threshold value is 6, b ...
 
-To find the best threshold for your model, we recommend experimenting with the `llm_int8_threshold` parameter in [`BitsAndBytesConfig`]:
+To find the best threshold for your model, experiment with the `llm_int8_threshold` parameter in [`BitsAndBytesConfig`].
```
This may be a good place for us to indicate that if you set the threshold to `0.0`, you'll get a significant improvement in inference speed, but with some potential accuracy trade-offs.

stevhliu committed:
feedback

stevhliu committed:
fix

stevhliu committed:
update

stevhliu commented:
Hey, I've addressed all the feedback so far. Please let me know if there is anything else, otherwise let's try and send it this week! 🤗

cc @ArthurZucker @LysandreJik

MekkCyber commented on docs/source/en/quantization/higgs.md:
````diff
@@ -16,11 +16,30 @@ rendered properly in your Markdown viewer.
+```
+
+</hfoption>
+<hfoption id="CUDA 11.8">
+
+```bash
+pip install flute-kernel -i https://flute-ai.github.io/whl/cu118
+```
+
+</hfoption>
````
can we add the 12.4 version ? installed with : `pip install flute-kernel -i https://flute-ai.github.io/whl/cu12.4`

MekkCyber commented on docs/source/en/quantization/quanto.md:
```diff
@@ -14,55 +14,56 @@ rendered properly in your Markdown viewer.
+# Optimum Quanto
 
-<Tip>
+[Quanto](https://github.com/huggingface/optimum-quanto) is a PyTorch quantization backend for [Optimum](https://huggingface.co/docs/optimum/index). It features linear quantization for weights (float8, int8, int4, int2) with accuracy very similar to full-precision models. Quanto is compatible with with any model modality and device, making it simple to use regardless of hardware.
```
tiny typo in the last sentence, "with" repeated twice

MekkCyber commented on docs/source/en/quantization/quanto.md:
```diff
@@ -14,55 +14,56 @@ rendered properly in your Markdown viewer.
-- device agnostic (e.g CUDA,XPU,MPS,CPU)
-- compatibility with `torch.compile`
-- easy to add custom kernel for specific device
-- supports quantization aware training
-<!-- Add link to the blogpost -->
-
```
I couldn’t find this anywhere as it seems to be deleted. Apologies if I’m mistaken, but it would be great to have it added somewhere.

stevhliu replied:
This should be added to the end of the doc under the "Resources" section :)

MekkCyber commented:
Thanks @stevhliu for this huge work ! I only added some few nits

stevhliu committed:
feedback

stevhliu committed:
updates

ArthurZucker reviewed (approved):
🤗 thanks for addressing and updating !

LysandreJik reviewed (approved):
Love it! Thanks @stevhliu, let's ship it!

stevhliu committed:
update

Ninjadeveloper007 commented:
Ok guys let me the self taught developer have it roast my ass! just provide much needed mentorship afterwards?!🫣😬🥸, i 
## **luv u guys?**

🥹
